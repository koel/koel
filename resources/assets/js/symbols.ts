import { DeepReadonly, InjectionKey, Ref } from 'vue'
import DialogBox from '@/components/ui/DialogBox.vue'
import MessageToaster from '@/components/ui/MessageToaster.vue'
import Router from '@/router'

export interface ReadonlyInjectionKey<T> extends InjectionKey<[Readonly<T> | DeepReadonly<T>, Closure]> {
}

export const RouterKey: InjectionKey<Router> = Symbol('Router')
export const ScreenNameKey: ReadonlyInjectionKey<ScreenName> = Symbol('ScreenName')

export const DialogBoxKey: InjectionKey<Ref<InstanceType<typeof DialogBox>>> = Symbol('DialogBox')
export const MessageToasterKey: InjectionKey<Ref<InstanceType<typeof MessageToaster>>> = Symbol('MessageToaster')

export const SongsKey: ReadonlyInjectionKey<Ref<Song[]>> | InjectionKey<Ref<Song[]>> = Symbol('Songs')
export const CurrentSongKey: InjectionKey<Ref<Song | null>> = Symbol('CurrentSong')
export const SelectedSongsKey: ReadonlyInjectionKey<Ref<Song[]>> = Symbol('SelectedSongs')
export const SongListConfigKey: ReadonlyInjectionKey<Partial<SongListConfig>> = Symbol('SongListConfig')
export const SongListSortFieldKey: ReadonlyInjectionKey<Ref<SongListSortField>> = Symbol('SongListSortField')
export const SongListSortOrderKey: ReadonlyInjectionKey<Ref<SortOrder>> = Symbol('SongListSortOrder')

export const EditSongFormInitialTabKey: ReadonlyInjectionKey<Ref<EditSongFormTabName>> = Symbol('EditSongFormInitialTab')

export const PlaylistKey: ReadonlyInjectionKey<Ref<Playlist>> = Symbol('Playlist')
export const PlaylistFolderKey: ReadonlyInjectionKey<Ref<PlaylistFolder>> = Symbol('PlaylistFolder')
export const UserKey: ReadonlyInjectionKey<Ref<User>> = Symbol('User')
