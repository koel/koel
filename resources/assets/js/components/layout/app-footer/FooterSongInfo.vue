<template>
  <div class="song-info" :class="{ playing: song?.playback_state === 'Playing' }">
    <span :style="{ backgroundImage: `url('${cover}')` }" class="album-thumb"/>
    <div class="meta" v-if="song">
      <h3 class="title">{{ song.title }}</h3>
      <a :href="`/#/artist/${song.artist_id}`" class="artist">{{ song.artist_name }}</a>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed, ref } from 'vue'
import { defaultCover, requireInjection } from '@/utils'
import { CurrentSongKey } from '@/symbols'

const song = requireInjection(CurrentSongKey, ref(null))

const cover = computed(() => song.value?.album_cover || defaultCover)
</script>

<style lang="scss" scoped>
.song-info {
  padding: 0 1.5rem;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  width: 320px;
  gap: 1rem;

  @media screen and (max-width: 768px) {
    width: 84px;
  }

  .album-thumb {
    display: block;
    height: 75%;
    aspect-ratio: 1;
    border-radius: 50%;
    background-size: cover;

    @media screen and (max-width: 768px) {
      height: 55%;
    }
  }

  .meta {
    overflow: hidden;

    > * {
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
    }

    @media screen and (max-width: 768px) {
      display: none;
    }
  }

  .artist {
    display: block;
    font-size: .9rem;
  }

  &.playing .album-thumb {
    animation: spin 30s linear infinite;

    @media (prefers-reduced-motion) {
      animation: none;
    }
  }
}

@keyframes spin {
  100% {
    transform: rotate(360deg);
  }
}
</style>
