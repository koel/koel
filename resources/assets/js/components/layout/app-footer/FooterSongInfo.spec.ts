import { expect, it } from 'vitest'
import factory from '@/__tests__/factory'
import UnitTestCase from '@/__tests__/UnitTestCase'
import FooterSongInfo from './FooterSongInfo.vue'
import { ref } from 'vue'
import { CurrentSongKey } from '@/symbols'

new class extends UnitTestCase {
  protected test () {
    it('renders with no current song', () => expect(this.render(FooterSongInfo).html()).toMatchSnapshot())

    it('renders with current song', () => {
      const song = factory<Song>('song', {
        title: 'Fahrstuhl zum Mond',
        album_cover: 'https://via.placeholder.com/150',
        playback_state: 'Playing',
        artist_id: 10,
        artist_name: 'Led Zeppelin'
      })

      expect(this.render(FooterSongInfo, {
        global: {
          provide: {
            [CurrentSongKey]: ref(song)
          }
        }
      }).html()).toMatchSnapshot()
    })
  }
}
