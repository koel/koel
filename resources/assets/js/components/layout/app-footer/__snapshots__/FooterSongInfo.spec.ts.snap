// Vitest Snapshot v1

exports[`renders with current song 1`] = `
<div class="song-info playing" data-v-91ed60f7=""><span style="background-image: url(https://via.placeholder.com/150);" class="album-thumb" data-v-91ed60f7=""></span>
  <div class="meta" data-v-91ed60f7="">
    <h3 class="title" data-v-91ed60f7="">Fahrstuhl zum Mond</h3><a href="/#/artist/10" class="artist" data-v-91ed60f7="">Led Zeppelin</a>
  </div>
</div>
`;

exports[`renders with no current song 1`] = `
<div class="song-info" data-v-91ed60f7=""><span style="background-image: url(undefined/resources/assets/img/covers/default.svg);" class="album-thumb" data-v-91ed60f7=""></span>
  <!--v-if-->
</div>
`;
